integration-tests/tests/src/tests/sync/flexible.ts

```diff
@@ -396,6 +398,85 @@ describe.skipIf(environment.missingServer, "Flexible sync", function () {
+        expect(error.code).to.equal(231);
+        expect(error.isFatal).to.be.false;
+        expect(error.message).to.contain(
+          "Client attempted a write that is outside of permissions or query filters; it has been reverted",
+        );
+
+        if (!(error instanceof CompensatingWriteError)) {
+          throw new Error("Expected a CompensatingWriteError");
+        }
+
+        expect(error.writes.length).to.equal(3);
+
+        const compensatingWrites = error.writes.sort((a, b) =>
+          (a.primaryKey as BSON.ObjectId).toString().localeCompare((b.primaryKey as BSON.ObjectId).toString()),
+        );
+
+        expect((compensatingWrites[0].primaryKey as BSON.ObjectId).equals(person1Id)).to.be.true;
+        expect((compensatingWrites[1].primaryKey as BSON.ObjectId).equals(person2Id)).to.be.true;
+        expect((compensatingWrites[2].primaryKey as BSON.ObjectId).equals(dogId)).to.be.true;
+
+        expect(compensatingWrites[0].objectName).to.equal(FlexiblePersonSchema.name);
+        expect(compensatingWrites[1].objectName).to.equal(FlexiblePersonSchema.name);
+        expect(compensatingWrites[2].objectName).to.equal(DogSchema.name);
+
+        expect(compensatingWrites[0].reason).to.contain("object is outside of the current query view");
+        expect(compensatingWrites[1].reason).to.contain("object is outside of the current query view");
+        expect(compensatingWrites[2].reason).to.contain("object is outside of the current query view");
+
+        callbackHandle.resolve();
+      };
+
+      const realm = await Realm.open({
+        schema: [FlexiblePersonSchema, DogSchema],
+        sync: {
+          flexible: true,
+          user: this.user,
+          onError: errorCallback,
+        },
+      });
+
+      await realm.subscriptions.update((mutableSubs) => {
+        mutableSubs.add(realm.objects(FlexiblePersonSchema.name).filtered("age < 30"));
+        mutableSubs.add(realm.objects(DogSchema.name).filtered("age > 5"));
+      });
+
+      realm.write(() => {
+        //Outside subscriptions
+        const tom = realm.create<IPerson>(FlexiblePersonSchema.name, {
+          _id: person1Id,
+          name: "Tom",
+          age: 36,
+        });
+        realm.create<IPerson>(FlexiblePersonSchema.name, { _id: person2Id, name: "Maria", age: 44 });
+        realm.create<IDog>(DogSchema.name, { _id: dogId, name: "Puppy", age: 1, owner: tom });
+
+        //Inside subscriptions
+        const luigi = realm.create<IPerson>(FlexiblePersonSchema.name, {
+          _id: new BSON.ObjectId(),
+          name: "Luigi",
+          age: 20,
+        });
+        realm.create<IPerson>(FlexiblePersonSchema.name, { _id: new BSON.ObjectId(), name: "Mario", age: 22 });
+        realm.create<IDog>(DogSchema.name, { _id: new BSON.ObjectId(), name: "Oldy", age: 6, owner: luigi });
+      });
+
+      await realm.syncSession?.uploadAllLocalChanges();
+      await callbackHandle;
```

Comment:
Looking at this, we should probably change this API to return something that can actually be awaited directly 🤔  but you have to await the `promise` property on the handle instead of the handle itself.

```suggestion
      await callbackHandle.promise;
```